### ISSUE TITLE
[Question] Flask and dependency_injector

### ISSUE THREAD

**Issue by picaso** [ets-labs/python-dependency-injector#159]
First off; thank you for much for such a wonderful tool. I enjoy using this a lot. I have a question that I hope you can educate me on.

I am writing a client-service which can handle rest calls.
When I make a request, the client service would perform some i/o on the client's machine folder (I write to a folder) so it feeds data to some internal process (proprietary legacy services) and the client-service scans another folder for a response from said legacy service and takes action (make a rest call posting what was found).
I am using watchdog observer for the directory/folder scan.

**client-service:** I have a controller (blueprint) which delegates to a service and I made them threadSafeSingletons. I injected the observer into the service as a threadSafeSingleton which starts scanning for changes on service instantiation.

I am making it a threadSafeSingleton because the observer runs on a separate thread and I want it to be long running, and any new instances scanning the same folder throws errors. Seeing as flask creates a new context on every request and assuming I run this using wsgi, is the threadsafe singleton going to be fine?
fyi: I have not tested this on some wsgi server yet. Thanks!

**Comment by rmk135** [comment]
Hi @picaso ,

Thanks a lot for your feedback, it gives energy to move ahead with things like this :)

> is the threadsafe singleton going to be fine?

- Yes, absolutely. ``ThreadSafeSingleton`` provider gives you 100% confidence that you will have one and the only instance of object in multithreaded application. And, yes, I do think that usage of ``ThreadSafeSingleton`` provider is justified for this case.

In addition, I would say that I see architecture like you describe not for a first time. Flask application + thread that observes system waiting for some events or interacting with some system software - is like a classic way of giving REST API for these system components.

Thanks,
Roman

**Comment by rmk135** [comment]
I close the issue, but, please, feel free to reopen it if needed ;)

**Comment by c650** [comment]
Hey there,

I found this while exploring an issue. I'm running a Connexion (Flask+Swagger) app and I used `@inject` to provide some `ThreadSafeSingleton`s. Sometimes they were `None` (due to synchronisation?), but it seems to work if I hide them behind another resource that is a `providers.Factory`. Is this intended?

I was expecting that the provider would block until it could get the `ThreadSafeSingleton`, but evidently not.
